### PR TITLE
Create flow for VZ narrative OCR & diagram extraction

### PR DESCRIPTION
**Pull request by frankhereford**
## TLDR

This is a flow which can orchestrate / schedule the CR3 inspection task which OCR's the crash narrative and extracts the crash diagram. It executes the existing python script which is unchanged. 

This exercise is also intended to create boilerplate code for wrapping an **arbitrary shell command**, such as a stand-alone python script, into a prefect flow. ✨

## Environment

This work is intended to be used in the dockerized prefect environment as found in the [vz-crash-import](https://github.com/cityofaustin/atd-prefect/pull/9) branch and this PR requests to merge data into that branch.

## Notes

This PR intends to close https://github.com/cityofaustin/atd-data-tech/issues/9419. 

## Testing

Defining how to test an ETL / integration is not as clear to me as we might have for a moped PR. Please feel free to do any, all, or other things based on your review of the contents of the PR. I'm very open to feedback on how to propose testing.

- [ ] Review the code
- [ ] Spin up the environment, with consideration to the KV store you're going to use.
- [ ] Cause the Flow to be executed via the Prefect web-interface or via local execution,`flow.run()`
- [ ] Observe that diagrams are extracted and OCR'd narratives are stored

### CODE REVIEW

File: flows/vision-zero/Dockerfile [hunk 1]
```diff
@@ -5,10 +5,11 @@ RUN apt-get update && \
     apt-get install -yq tzdata && \
     ln -fs /usr/share/zoneinfo/America/Chicago /etc/localtime && \
     dpkg-reconfigure -f noninteractive tzdata
-RUN apt upgrade -y
-RUN apt install -y vim magic-wormhole aptitude python3-pip docker.io rsync p7zip-full black
+RUN apt-get upgrade -y
+RUN apt-get install -y vim magic-wormhole aptitude python3-pip docker.io rsync p7zip-full black poppler-utils tesseract-ocr
 RUN pip install prefect
 RUN pip install sysrsync GitPython boto3
+RUN pip install pdf2image pytesseract
```
Comment (frankhereford) [inline flows/vision-zero/Dockerfile:12]: Adding some system tools and python wrappers around them. Poppler provides a tool to rasterize PDFs and tesseract is the OCR utility.

File: flows/vision-zero/cr3_ocr_narrative_extract_diagram/cr3_ocr_narrative_extract_diagram.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    command = (
+        f"/usr/bin/python3 \
+        /root/cr3_ocr_narrative_extract_diagram/atd-airflow/dags/python_scripts/cr3_extract_diagram_ocr_narrative.py "
+        + op_mode
+        + f"-vd --update-narrative --update-timestamp \
+        --cr3-source {OCR_CR3_SOURCE_BUCKET} {OCR_CR3_SOURCE_PATH} \
+        --save-diagram-s3 {OCR_DIAGRAM_TARGET_BUCKET} {OCR_DIAGRAM_TARGET_PATH}"
+    )
+
+    stream = task(command=command)
```
Comment (frankhereford) [inline flows/vision-zero/cr3_ocr_narrative_extract_diagram/cr3_ocr_narrative_extract_diagram.py:63]: This flow can be viewed as boilerplate to prefect-ize a single shell command, such as a stand-alone python script.

File: flows/vision-zero/make_kv_store.py [hunk 3]
```diff
@@ -17,6 +17,14 @@
...
     "AWS_CSV_ARCHIVE_PATH_PRODUCTION": os.getenv("AWS_CSV_ARCHIVE_PATH_PRODUCTION"),
     "AWS_CSV_ARCHIVE_PATH_STAGING": os.getenv("AWS_CSV_ARCHIVE_PATH_STAGING"),
+    "PRODUCTION_GRAPHQL_ENDPOINT": os.getenv("PRODUCTION_GRAPHQL_ENDPOINT"),
+    "PRODUCTION_GRAPHQL_ENDPOINT_ACCESS_KEY": os.getenv("PRODUCTION_GRAPHQL_ENDPOINT_ACCESS_KEY"),
+    "OCR_CR3_SOURCE_BUCKET": os.getenv("OCR_CR3_SOURCE_BUCKET"),
+    "OCR_CR3_SOURCE_PATH": os.getenv("OCR_CR3_SOURCE_PATH"),
+    "OCR_DIAGRAM_TARGET_BUCKET": os.getenv("OCR_DIAGRAM_TARGET_BUCKET"),
+    "OCR_DIAGRAM_TARGET_PATH": os.getenv("OCR_DIAGRAM_TARGET_PATH"),
+    "OCR_BATCH_SIZE": os.getenv("OCR_BATCH_SIZE"),
+    "OCR_SINGLE_CRASH": os.getenv("OCR_SINGLE_CRASH"),
```
Comment (frankhereford) [inline flows/vision-zero/make_kv_store.py:27]: Bringing more configuration values out of the Airflow JSON blob and into the top level JSON blob as found in the prefect KV store.

File: flows/vision-zero/make_kv_store.py [hunk 4]
```diff
@@ -17,6 +17,14 @@
     "AWS_CSV_ARCHIVE_BUCKET_NAME": os.getenv("AWS_CSV_ARCHIVE_BUCKET_NAME"),
```
Comment (frankhereford) [inline flows/vision-zero/make_kv_store.py:17]: This file was moved up to the project level folder as it is intended to create a KV store for _all_ VZ flows, not just one.

File: flows/vision-zero/cr3_ocr_narrative_extract_diagram/cr3_ocr_narrative_extract_diagram.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    logger.info(stream)
+
+#flow.run()
+flow.register(project_name="vision-zero")
```
Comment (chiaberry) [inline flows/vision-zero/cr3_ocr_narrative_extract_diagram/cr3_ocr_narrative_extract_diagram.py:68]: After registering a flow like this, does one then need to go into the UI to trigger the scheduled runs? 

Also, do you think adding an "idempotency_key" would be a good idea? https://docs.prefect.io/orchestration/concepts/flows.html#core-client 
Or would the `register` command be commented out and `flow.run()` be used after PR approval?

Comment (frankhereford) [reply flows/vision-zero/cr3_ocr_narrative_extract_diagram/cr3_ocr_narrative_extract_diagram.py:68]: Re scheduling:

Great catch! I do envision that the scheduling being done via the Prefect UI. The execution of the flow to register gets it queued up to be scheduled, and then the scheduling is contained in the UI, not dictated by code that may or may not be up to date. Seeking a single source of truth.

Comment (frankhereford) [reply flows/vision-zero/cr3_ocr_narrative_extract_diagram/cr3_ocr_narrative_extract_diagram.py:68]: Re the `indempotency_key` argument to the registration function:

TIL: indempotent - maths not changed in value following multiplication by itself. 

I think the jury is still out on if this will be needed. When I put this PR together, I was imagining the programmer being in charge of when things get registered and the agent getting rebuilt etc. As it was, I would just want to count on the developer to "do the right thing", but it was you and my discussion from the other day that prompted me to reevaluate that. After that conversation and with an issue put on deck for me by product, I am going to seek automation via GitHub actions for deployment related tasks. It'll be this work which will inform the need of the `indempotency_key`.

**Review by chiaberry — APPROVED** [review]
I had questions around how registering flows would work, but that will be addressed in this issue: https://github.com/cityofaustin/atd-data-tech/issues/9612
so 👍🏼

File: .gitmodules [hunk 6]
```diff
@@ -1,3 +1,6 @@
 [submodule "flows/vision_zero/atd-vz-data"]
 	path = flows/vision-zero/cris_import/atd-vz-data
 	url = https://github.com/cityofaustin/atd-vz-data
+[submodule "flows/vision-zero/cr3_ocr_narrative_extract_diagram/atd-airflow"]
+	path = flows/vision-zero/cr3_ocr_narrative_extract_diagram/atd-airflow
+	url = https://github.com/cityofaustin/atd-airflow.git
```
[the remark this answers is not in the document]
Comment (mddilley) [reply .gitmodules:6]: I'm still getting a hang of submodules when developing in the dockerized-agent system so I think that documenting that a little would be a good step to getting everyone on the same page. I know this is still in progress, and I am happy to help with this effort. 

Totally not needed to get this merged, and I think getting some hands-on experience like we have with the Moped test deployment will be an important part of knowing how to bridge between this repo and other resources.

**Review by mddilley — APPROVED** [review]
This has really helped me wrap my mind around the ShellTasks that are in the Moped API code and how I may be able to break down that part of the flow into smaller pieces. Glad that I read through this again as I try to Prefect-ize my thought process. 🙏 

All in all - this looks great! I think it is a good way to get an existing script running in Prefect and also a nice path for future scripts that doesn't require a lot of spin-up.

**Review by patrickm02L — APPROVED** [review]
Can't wait to fire this up! 🔥
